File: plugin/src/testKit/java/com/mobilesolutionworks/gradle/jacoco/jupiter/GradleRunnerWrapper.java
package com.mobilesolutionworks.gradle.jacoco.jupiter;

import org.gradle.testkit.runner.BuildResult;
import org.gradle.testkit.runner.GradleRunner;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class GradleRunnerWrapper {

    public final GradleRunner runner;

    public boolean output = false;

    private List<String> arguments = new ArrayList<>();

    public GradleRunnerWrapper(GradleRunner runner) {
        this.runner = runner;
    }

    void setArgs(String... args) {
        arguments.addAll(Arrays.asList(args));
    }

    private void apply() {
        if (output) {
            runner.forwardOutput();
        }

        runner.withArguments(arguments);
    }

    public BuildResult build() {
        apply();
        return runner.withPluginClasspath().build();
    }

    public File getRoot() {
        return runner.getProjectDir();
    }
}

File: plugin/src/testKitKotlin/java/com/mobilesolutionworks/gradle/jacoco/jupiter/GradleRunnerWrapper.kt
package com.mobilesolutionworks.gradle.jacoco.jupiter

import org.gradle.testkit.runner.BuildResult
import org.gradle.testkit.runner.GradleRunner
import java.io.File

class GradleRunnerWrapper(internal val runner: GradleRunner) {

    val root: File
        get() = runner.projectDir

    var output = false

    private var arguments = mutableListOf<String>()

    init {
    }

    fun args(vararg args: String) {
        arguments.addAll(args)
    }

    fun configure(closure: GradleRunnerWrapper.() -> Unit): GradleRunnerWrapper {
        closure()
        return this
    }

    private fun apply() {
        if (output) {
            runner.forwardOutput()
        }

        runner.withArguments(arguments)
    }

    fun build(closure: BuildResult.() -> Unit) {
        apply()
        runner.withPluginClasspath().build().closure()
    }
}
